Write docs on how to use Empire.

brianz opened the pull request:
I wrote some docs about using Empire.  Of course, this is a view as I understand Empire now.  If I got anything wrong please correct me and I'll update this!

Explains the biggest points on:

- how to structure your application
- Procfiles
- Routing rules
- Different application types, web vs. non-web

TODO - Write some stuff about using environment variables in the app.  Perhaps outside the scope of
Empire, but still relevant.

phobologic commented on docs/deploying_an_application.md:
````diff
@@ -0,0 +1,109 @@
+```
+
+
+## Empire applicaiton types
````
spelling

phobologic commented on docs/deploying_an_application.md:
```diff
@@ -0,0 +1,109 @@
+
+## Procfile
+
+Empire was modeled after the Heroku API.  As such, many concepts and commands are similar.  One concept which has bled over and which is a requirement is that of a [Procfile][procfiles].  For your application to actually work, you must include a Procfile at the root level of your appplication.
```
So it turns out that a Procfile *isn't* necessarily required.  Empire tries to do the right thing when told to deploy an app without a Procfile.  Still, a Procfile is a lot better than hoping Empire gets it right.  This is probably fine here.

phobologic commented on docs/deploying_an_application.md:
```diff
@@ -0,0 +1,109 @@
+For the impatient, there are two absolutes:
+
+- Your must create a `Procfile` which defines how to run your application
+- If you application answers requests over the network, it must listen on port 8080, or better, on the `$PORT` environment variable.
```
Might be worth mentioning that we really, really suggest you use $PORT if you can.  The reason being that it will allow us to change it later without breaking the app.  I don't see a good reason why we might change it, but it'd be cool to be able to.

ejholmes replied:
s/you/your

benjamin-guillet commented on docs/deploying_an_application.md:
```diff
@@ -0,0 +1,109 @@
+# Empire :: Deploying an Application
+
+Before using Empire successfully it's important to understand it's design and architecture. It's suggested you first read the `Features & Design Goals document`.
```
`its design`

benjamin-guillet replied:
`its design`

benjamin-guillet commented on docs/deploying_an_application.md:
```diff
@@ -0,0 +1,109 @@
+Before using Empire successfully it's important to understand it's design and architecture. It's suggested you first read the `Features & Design Goals document`.
+
+There are a few key things which your applications, and overall architecture, will need to keep in mind in order to be successful with Empire.  Below is an exhaustive set of criteria your applications must meet in order to work with Empire.  It's not that long and by no means locks you in to Empire.  Structing your application in this way provides many benefits, only one of which is being able to ...
+
```
Spelling: structuring :)

benjamin-guillet commented on docs/deploying_an_application.md:
````diff
@@ -0,0 +1,109 @@
+web: python manage.py runserver
+```
+
+For web applications, Empire does a few things.  There are multiple layers of routing, none of which are extrelemy complex.
````
extremely :)

benjamin-guillet commented on docs/deploying_an_application.md:
```diff
@@ -0,0 +1,109 @@
+
+- Empire creates an internal HostedZone CNAME record for `awesome-app`.  This CNAME points at am ELB which Empire also creates, specifically for this application.  If you `emp deploy mycompany/another-app`, yet *another* CNAME and ELB would be created for `another-app`.
+- The ELB created is managed by empire. As you scale your application *up*, instances running your application are placed into the ELB.  Likewise, as you scale your application *down*, instances are removed from the ELB. The ELB listens on port 80 and maps to a random port between 9000 and 10000 on the minion instances running your application.
+- The ELB runs a health check to determin whether your application is healthy. It will simply perform a tcp `ping` to your application...if your app doesn't respond, you will end up in a state where there are no healthy instances behind the ELB.
```
determine

phobologic commented on docs/deploying_an_application.md:
````diff
@@ -0,0 +1,109 @@
+The routing to your application is handled as such inside of the VPC:
+
+```
+http://awesome-app --> ELB to minions (port 80 to [9000-10000]) --> Host port [9000-10000] -> container port 8080
````
I feel like this could be a little more clear.  Maybe:

```
http://awesome-app -> ELB (port 80) -> Minion (port [9000-10000]) -> container port $PORT (8080)
```

??

benjamin-guillet replied:
I agree with @phobologic on that one :)

ejholmes commented on docs/deploying_an_application.md:
```diff
@@ -0,0 +1,109 @@
+
+For the impatient, there are two absolutes:
+
+- Your must create a `Procfile` which defines how to run your application
```
s/Your/You

mhahn commented on docs/deploying_an_application.md:
```diff
@@ -0,0 +1,109 @@
+
+### Non-web processes
+
+Let's start with non-web processes first since they're much simpler.  If your service doens't expose a port it doesn't need to be discovered by any other system.  With that, all you need to do it deploy your application and let Empire manage it. The act of discovering *other* services is up to you (database, caches, etc.)
```
*doesn't

benjamin-guillet commented on docs/deploying_an_application.md:
```diff
@@ -0,0 +1,109 @@
+
+
+There are a few other environment variable which Empire will set for your running container, but `PORT` is the most imoportant for now.  There are various ways to get your application listening on `$PORT` if you prefer not to hardcode `8080`. It's suggested to use `$PORT` rather than 8080.  One such way is to run your application from a shell script, which is in turn used in your `Procfile`
+
```
spelling: important

benjamin-guillet commented:
this looks good :)

ejholmes commented on docs/deploying_an_application.md:
```diff
@@ -0,0 +1,109 @@
+
+There are various things going on...let's break them down:
+
+- Empire creates an internal HostedZone CNAME record for `awesome-app`.  This CNAME points at am ELB which Empire also creates, specifically for this application.  If you `emp deploy mycompany/another-app`, yet *another* CNAME and ELB would be created for `another-app`.
```
s/at am/at an

ejholmes commented on docs/deploying_an_application.md:
```diff
@@ -0,0 +1,109 @@
+
+
+
+There are a few other environment variable which Empire will set for your running container, but `PORT` is the most imoportant for now.  There are various ways to get your application listening on `$PORT` if you prefer not to hardcode `8080`. It's suggested to use `$PORT` rather than 8080.  One such way is to run your application from a shell script, which is in turn used in your `Procfile`
```
s/variable/variables

ejholmes commented:
This is awesome! Thanks @brianz!

rgabo commented on docs/deploying_an_application.md:
```diff
@@ -0,0 +1,109 @@
+There are various things going on...let's break them down:
+
+- Empire creates an internal HostedZone CNAME record for `awesome-app`.  This CNAME points at am ELB which Empire also creates, specifically for this application.  If you `emp deploy mycompany/another-app`, yet *another* CNAME and ELB would be created for `another-app`.
+- The ELB created is managed by empire. As you scale your application *up*, instances running your application are placed into the ELB.  Likewise, as you scale your application *down*, instances are removed from the ELB. The ELB listens on port 80 and maps to a random port between 9000 and 10000 on the minion instances running your application.
```
It's easier to follow the text if Empire is always capitalized correctly :smile:

brianz commented:
Thanks for the proof reads @benjamin-guillet @ejholmes @rgabo @phobologic @mhahn 

Corrected spelling mistakes and updated some other parts based on feedback.  Hopefully the routing ASCII diagram is a bit more clear now.

phobologic commented on docs/deploying_an_application.md:
```diff
@@ -0,0 +1,123 @@
+There are various things going on...let's break them down:
+
+- Empire creates an internal HostedZone CNAME record for `awesome-app`.  This CNAME points at an ELB which Empire also creates, specifically for this application.  If you `emp deploy mycompany/another-app`, yet *another* CNAME and ELB would be created for `another-app`.
+- The ELB created is managed by Empire. As you scale your application *up*, instances running your application are placed into the ELB.  Likewise, as you scale your application *down*, instances are removed from the ELB. The ELB listens on port 80 and maps to a random port between 9000 and 10000 on the minion instances running your application.
```
hmm - random isn't quite right, though it's probably safe enough for this.  Might be worth mentioning that each app gets a port assigned to it - so you won't have two apps with the same port.  We actually just pick the next port avaiable, incrementing from 9000-10000.  Not sure of an easy way to explain that.

phobologic commented:
:+1: Looks good to me.

ejholmes commented:
:+1:
